### PR TITLE
[GPoS] Support Stage-2 stake limit calculation logic

### LINKED ISSUE #551: [GPoS] Stage 2 Stake Limit Calculation

### PR DESCRIPTION
**Pull request by MyronFanQiu**
Fix #551

### CODE REVIEW

File: cstrml/staking/src/lib.rs [hunk 1]
```diff
@@ -1528,31 +1537,51 @@ impl<T: Config> Module<T> {
...
-            storage_stakes.try_into().ok().unwrap()
-        } else {
-            (u64::max_value() as u128).try_into().ok().unwrap()
+    pub fn add_extra_ratio_according_to_effective_staking(total_issuance: BalanceOf<T>) -> Perbill {
+        let to_num =
+            |b: BalanceOf<T>| <T::CurrencyToVote as Convert<BalanceOf<T>, u64>>::convert(b);
+        // Multiple the percent by four, maximum value is 100%
+        let mul_four = |percent: Perbill| {
+            let mut init_percent = Perbill::zero();
+            for _ in 0..4 {
```
Comment (zikunfan) [inline cstrml/staking/src/lib.rs:1571]: Put this `4` and that `25` into config

File: cstrml/staking/src/lib.rs [hunk 2]
```diff
@@ -1528,31 +1537,51 @@ impl<T: Config> Module<T> {
...
 
-        // Now, we apply directly mapping algorithm for the early stage:
-        // 1. Maxwell 1.0: 1 terabytes -> 80,000 CRUs
-        // 2. Olympus 1.0: 1 terabytes -> 30 CRUs(tmp)
-        // ps: we treat 1 terabytes as 1_000_000_000_000 for make `mapping_ratio = 1`
-        if let Some(storage_stakes) = own_workloads.checked_mul(T::SPowerRatio::get()) {
-            storage_stakes.try_into().ok().unwrap()
-        } else {
-            (u64::max_value() as u128).try_into().ok().unwrap()
+    pub fn add_extra_ratio_according_to_effective_staking(total_issuance: BalanceOf<T>) -> Perbill {
```
Comment (zikunfan) [inline cstrml/staking/src/lib.rs:1565]: Just named `extra_limit_ratio` cause there is no `add` operations here

File: cstrml/staking/src/lib.rs [hunk 3]
```diff
@@ -2437,12 +2466,37 @@ impl<T: Config> historical::SessionManager<T::AccountId, Exposure<T::AccountId,
...
-            let v_own_workload = workload_map.get(&v_stash).unwrap_or(&0u128);
-            Self::upsert_stake_limit(
-                &v_stash,
-                Self::stake_limit_of(*v_own_workload, total_workload),
-            );
+        let total_issuance = T::Currency::total_issuance();
+        let stake_limit_ratio = Self::stake_limit_ratio();
+        // If effective staking ratio is smaller than some value, we should increase the total stake limit
+        let extra_stake_limit_ratio = Self::add_extra_ratio_according_to_effective_staking(total_issuance.clone());
+        let total_stake_limit = TryInto::<u128>::try_into((stake_limit_ratio * total_issuance).saturating_add(extra_stake_limit_ratio * total_issuance))
```
Comment (zikunfan) [inline cstrml/staking/src/lib.rs:2473]: Add some comments on this value may `> total_issuance`

File: cstrml/staking/src/lib.rs [hunk 4]
```diff
@@ -2437,12 +2466,37 @@ impl<T: Config> historical::SessionManager<T::AccountId, Exposure<T::AccountId,
...
+            .ok()
+            .unwrap();
+        // Judge if we're in stage one or stage two
+        let is_stage_one = total_workload.saturating_mul(T::SPowerRatio::get()) < total_stake_limit;
```
Comment (zikunfan) [inline cstrml/staking/src/lib.rs:2477]: I don't this this judging should be appeared in here, abstract this juding into a function

File: cstrml/staking/src/lib.rs [hunk 5]
```diff
@@ -2437,12 +2466,37 @@ impl<T: Config> historical::SessionManager<T::AccountId, Exposure<T::AccountId,
...
+
+        if is_stage_one {
```
Comment (zikunfan) [inline cstrml/staking/src/lib.rs:2479]: `for...each` then get stake limit with a unify entrypoint

File: cstrml/staking/src/lib.rs [hunk 6]
```diff
@@ -2437,12 +2466,37 @@ impl<T: Config> historical::SessionManager<T::AccountId, Exposure<T::AccountId,
...
+            for (v_stash, _) in <Validators<T>>::iter() {
+                let v_own_workload = workload_map.get(&v_stash).unwrap_or(&0u128);
+                Self::upsert_stake_limit(
+                    &v_stash,
+                    Self::stage_one_stake_limit_of(*v_own_workload),
+                );
+            }
+        } else {
+            // Decrease the precision to kb to avoid overflow
+            let total_workload_in_kb = total_workload / 1024;
```
Comment (zikunfan) [inline cstrml/staking/src/lib.rs:2490]: Abstract it as a util function, byte_to_kilobyte

**Review by zikunfan — CHANGES_REQUESTED** [review]

File: cstrml/staking/src/lib.rs (hunk 1 above)
Comment (MyronFanQiu) [reply cstrml/staking/src/lib.rs:1571]: leave the 4 since it should be a function instead of a value.

**Review by zikunfan — APPROVED** [review]

**Pull request merged**